hare/mock_oracle.go

```diff
@@ -14,34 +17,72 @@ const (
 )
 
 type Rolacle interface {
-	Role(r uint32, sig Signature) Role
+	Role(sig Signature) Role
 }
 
 type MockHashOracle struct {
-	isLeaderTaken bool
+	clients    map[string]struct{}
+	comitySize int
+	mutex      sync.Mutex
 }
 
-func NewMockOracle() *MockHashOracle {
-	mock := &MockHashOracle{}
+// N is the expected comity size
+func NewMockHashOracle(expectedSize int, comitySize int) *MockHashOracle {
+	mock := new(MockHashOracle)
+	mock.clients = make(map[string]struct{}, expectedSize)
+	mock.comitySize = comitySize
 
 	return mock
 }
 
-func (mockOracle *MockHashOracle) Role(r uint32, proof Signature) Role {
+func (mock *MockHashOracle) Register(pubKey crypto.PublicKey) {
+	mock.mutex.Lock()
+	defer mock.mutex.Unlock()
+
+	if _, exist := mock.clients[pubKey.String()]; exist {
+		return
+	}
+
+	mock.clients[pubKey.String()] = struct{}{}
+}
+
+func (mock *MockHashOracle) Unregister(pubKey crypto.PublicKey) {
+	mock.mutex.Lock()
+	delete(mock.clients, pubKey.String())
+	mock.mutex.Unlock()
+}
+
+func (mock *MockHashOracle) calcThreshold(comitySize int) uint32 {
+	if len(mock.clients) == 0 {
+		panic("Called calcThreshold with 0 clients registered")
+	}
+
+	return uint32(uint64(comitySize) * uint64(math.MaxUint32) / uint64(len(mock.clients)))
+}
+
+func (mock *MockHashOracle) Role(proof Signature) Role {
 	if proof == nil {
+		log.Warning("Oracle query with proof=nil. Returning passive")
 		return Passive
 	}
 
-	data := binary.LittleEndian.Uint32(proof)
+	// calculate thresholds
+	threshLeader := mock.calcThreshold(3)               // expect 3 leaders
+	threshActive := mock.calcThreshold(mock.comitySize) // expect comitySize-3 actives
 
-	if data < 10 {
+	// calculate hash of proof
+	hash := fnv.New32()
```

Comment:
you should have a struct responsible for hashing, this struct should also return the max value (instead of the maxUint32 above). This way you'll make sure that no one changes the hash type without forgetting to update the max value

Reply:
Fixed